src/utils/src/String.c

```diff
@@ -449,4 +449,89 @@ STATUS tolowerupperstr(PCHAR pStr, UINT32 strLen, BOOL toUpper, PCHAR pConverted
     }
 
     return retStatus;
-}
\ No newline at end of file
+}
+
+// This is a string search using Rabin–Karp algorithm.
+// Running time complexity:
+//   * Average: Θ(n + m)
+//   * Worst  : Θ((n−m)m)
+// Memory complexity: O(1)
+PCHAR defaultStrnstr(PCHAR haystack, PCHAR needle, SIZE_T len)
+{
+    UINT32 prime = 16777619;
+    UINT32 windowHash = 0;
+    UINT32 needleHash = 0;
+    UINT32 i;
+    UINT32 square = prime, power = 1;
+    UINT32 haystackSize, needleSize;
+
+    if (needle == NULL) {
+        return haystack;
+    }
+
+    if (haystack == NULL) {
+        return NULL;
+    }
+
+    haystackSize = (UINT32) STRNLEN(haystack, len);
+    needleSize = (UINT32) STRLEN(needle);
+    if (needleSize > haystackSize) {
+        return NULL;
+    }
+
+    for (i = 0; i < needleSize; i++) {
+        windowHash = windowHash * prime + (UINT32) haystack[i];
+        needleHash = needleHash * prime + (UINT32) needle[i];
+    }
+
+    if (windowHash == needleHash && STRNCMP(haystack, needle, needleSize) == 0) {
+        return haystack;
+    }
+
+    // Precompute the largest power in O(log n). 
+    //
+    // This algorithm takes an advantage for the fact that a number that's multiplied 
+    // by itself will double the power.
+    //
+    // For example:
+    //    a ^ 4 = (a ^ 2) * (a ^ 2) = a * a * a * a * a * a * a * a 
```

Comment:
You have a ^ 8 in  a * a * a * a * a * a * a * a

Reply:
Oops, I can't count...